File: Leetcode/cpp/medium/198.cpp
/*
 * @lc app=leetcode.cn id=198 lang=cpp
 *
 * [198] 打家劫舍
 */

// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
class Solution1 {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> memo(n, -1);
        auto dfs = [&](this auto&& dfs, int i) -> int {
            if (i < 0) return 0;
            if (memo[i] != -1) return memo[i];
            return memo[i] = max(dfs(i - 2) + nums[i], dfs(i - 1));
        };
        return dfs(n - 1);
    }
};

class Solution2 {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> f(n + 2, 0);
        for (int i = 2; i <= n + 1; i++)
            f[i] = max(f[i - 2] + nums[i - 2], f[i - 1]);
        return f[n + 1];
    }
};

class Solution3 {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        int f0 = 0, f1 = 0;
        for (int x : nums)
            tie(f0, f1) = tuple(f1, max(f0 + x, f1));
        return f1;
    }
};

// using Solution = Solution1;
// using Solution = Solution2;
using Solution = Solution3;
// @lc code=end

File: Leetcode/rust/medium/198.rs
/*
 * @lc app=leetcode id=198 lang=rust
 *
 * [198] House Robber
 */


// @lcpr-template-start
struct Solution;
// @lcpr-template-end
// @lc code=start
struct Solution1;
impl Solution1 {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut memo = vec![-1; n];
        fn dfs(i: isize, nums: &[i32], memo: &mut [i32]) -> i32 {
            if i < 0 { return 0; }
            let idx = i as usize;
            if memo[idx] != -1 { return memo[idx]; }
            memo[idx] = std::cmp::max(dfs(i - 2, nums, memo) + nums[idx], dfs(i - 1, nums, memo));
            memo[idx]
        }
        dfs(n as isize - 1, &nums, &mut memo)
    }
}

struct Solution2;
impl Solution2 {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut f = vec![0; n + 2];
        for (i, &x) in nums.iter().enumerate() {
            f[i + 2] = std::cmp::max(f[i] + x, f[i + 1]);
        }
        f[n + 1]
    }
}

struct Solution3;
impl Solution3 {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let (mut f0, mut f1) = (0, 0);
        for x in nums {
            (f0, f1) = (f1, std::cmp::max(f0 + x, f1));
        }
        f1
    }
}

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        // Solution1::rob(nums)
        // Solution2::rob(nums)
        Solution3::rob(nums)
    }
}
// @lc code=end

